static/js/unread_ops.js

```diff
@@ -9,32 +13,69 @@ import * as notifications from "./notifications";
+        anchor: "first_unread",
+        include_anchor: true,
+        num_before: 0,
+        num_after: NUM_OF_MESSAGES_UPDATED_PER_BATCH,
+        op: "add",
+        flag: "read",
+    };
     channel.post({
-        url: "/json/mark_all_as_read",
-        success: () => {
-            // After marking all messages as read, we reload the browser.
-            // This is useful to avoid leaving ourselves deep in the past.
-            // This is also the currently intended behavior in case of partial success,
-            // (response code 200 with result "partially_completed")
-            // where the request times out after marking some messages as read,
-            // so we don't need to distinguish that scenario here.
-            // TODO: The frontend handling of partial success can be improved
-            // by re-running the request in a loop, while showing some status indicator
-            // to the user.
-            reload.initiate({
-                immediate: true,
-                save_pointer: false,
-                save_narrow: true,
-                save_compose: true,
-            });
+        url: "/json/messages/flags",
+        data: opts,
+        success(data) {
+            if (!data.found_newest) {
+                if (NUM_OF_MESSAGES_UPDATED_PER_BATCH < unread_count_before_marking_all_msgs_read) {
+                    messages_read_till_now += NUM_OF_MESSAGES_UPDATED_PER_BATCH;
+                    let percentage_updated =
+                        messages_read_till_now / unread_count_before_marking_all_msgs_read;
+                    percentage_updated = (percentage_updated * 100).toFixed(2);
+                    ui_report.loading(
+                        $t_html(
+                            {defaultMessage: "Marking messages as read...{percentage_updated}%"},
+                            {percentage_updated},
+                        ),
+                        $("#request-progress-status-banner"),
+                    );
+                } else {
+                    ui_report.loading(
+                        $t_html({defaultMessage: "Marking messages as read..."}),
+                        $("#request-progress-status-banner"),
+                    );
+                }
+                if (!loading_indicator_displayed) {
+                    loading.make_indicator(
+                        $("#request-progress-status-banner .loading-indicator"),
+                        {abs_positioned: true},
+                    );
+                    loading_indicator_displayed = true;
+                }
+                mark_all_as_read();
+            } else {
+                unread_count_before_marking_all_msgs_read = 0;
+                messages_read_till_now = 0;
+                loading_indicator_displayed = false;
+                ui_report.loading(
+                    $t_html({defaultMessage: "Successfully marked all messages as read!"}),
+                    $("#request-progress-status-banner"),
```

Comment:
For the string, let's just use:

> Success!

We'll only show it after the progress banner, so it should be pretty clear what it's referring to.